magma/mantle/queue.py

```diff
@@ -0,0 +1,65 @@
+"""
+Based on
+https://github.com/chipsalliance/chisel3/blob/master/src/main/scala/chisel3/util/Decoupled.scala
+(missing support for `flow` and `pipe` parameters).
+"""
+from magma.clock_io import ClockIO
+from magma.conversions import enable
+from magma.generator import Generator2
+from magma.interface import IO
+from magma.mantle.counter import Counter
+from magma.primitives.memory import Memory
+from magma.primitives.register import Register
+from magma.types.handshake import DeqIO, EnqIO
+from magma.t import Kind
+
+
+class Queue(Generator2):
+    def __init__(self, entries: int, T: Kind):
+        assert entries >= 0
+        self.io = IO(
+            # NOTE(leonardt): Note that the direction of enq/deq are flipped
+            # since this is a client (consumer)
+            enq=DeqIO[T],
+            deq=EnqIO[T]
+        ) + ClockIO()
+
+        # Data storage
```

Comment:
these individual comments don't do much since the variable names and function names are pretty useful. instead let's add one big comment at the top of this function that describes the overall implementation, e.g.

This queue is implemented with a register file (RAM). Write (enqueue) and Read (dequeue) counters point to positions in the register file. etc.

Reply:
I added a top level comment covering the architecture, I still think the inline comments are useful for someone learning magma, so I'll leave them in with the idea that we can point to this code as an example of a typical circuit.